gossip/src/crds_value.rs

```diff
@@ -490,84 +495,233 @@ impl Sanitize for NodeInstance {
     }
 }
 
-#[derive(Serialize, Deserialize, Clone, Default, PartialEq, Eq, AbiExample, Debug)]
+#[derive(Serialize, Deserialize, Clone, Debug, PartialEq, Eq, AbiExample, AbiEnumVisitor)]
+pub enum SlotsOffsets {
+    RunLengthEncoding(RunLengthEncoding),
+    RawOffsets(RawOffsets),
+}
+
+#[derive(Deserialize, Serialize, Clone, Debug, PartialEq, Eq, AbiExample)]
+pub struct RunLengthEncoding {
+    // For every u8 in encoded, the last bit is the encoded bit, the previous
+    // bits are the count of encoded bit. For example, 0xF1 means there are
+    // (0xF0 >> 1) number of one's (the last bit), so this byte encoded
+    // 120 consecutive 1's. Each byte can encode maximum of 127 consecutive 0's
+    // or 1's.
+    encoded: Vec<u8>,
```

Comment:
This seems wasteful and unnecessarily completed.
Simpler approach is that you compute runs assuming that it always starts with consecutive `0`s. So for example:
```
1110010000  => [0,3,2,1,4]
```
then you would use varint for encoding the resulting vector.
https://github.com/solana-labs/solana/blob/a851670d5/sdk/program/src/serde_varint.rs 

If you can only serialize into `max_...` many bytes, you can easily compute how many entries of RLE vector you can serialize because each entry takes `(most-significant-bit + 6)/ 7` bytes to encode.

Reply:
Changed.